app/models/spree/gateway/stripe_gateway.rb

```diff
@@ -80,45 +84,112 @@ def create_profile(payment)
 
     private
 
-    # In this gateway, what we call 'secret_key' is the 'login'
+    # In this gateway, what we call "secret_key" is the "login"
     def options
       options = super
-      options.merge(:login => preferred_secret_key)
+
+      options.merge(login: preferred_secret_key)
     end
 
     def options_for_purchase_or_auth(money, creditcard, gateway_options)
       options = {}
-      order = gateway_options[:originator].order
-      options[:description] = "#{order.store.name.first(15)}-#{order.visual_id}"
       options[:currency] = gateway_options[:currency]
 
-      if customer = creditcard.gateway_customer_profile_id
+      if gateway_options.dig(:originator) && gateway_options[:originator].order_id.present?
+        merch_order = Spree::Order.find_by(id: gateway_options[:originator][:order_id])
+
+        apply_level3_data!(options, merch_order)
+      end
+
+      if (customer = creditcard.gateway_customer_profile_id)
         options[:customer] = customer
       end
-      if token_or_card_id = creditcard.gateway_payment_profile_id
+
+      if (token_or_card_id = creditcard.gateway_payment_profile_id)
         # The Stripe ActiveMerchant gateway supports passing the token directly as the creditcard parameter
         # The Stripe ActiveMerchant gateway supports passing the customer_id and credit_card id
         # https://github.com/Shopify/active_merchant/issues/770
         creditcard = token_or_card_id
       end
+
       return money, creditcard, options
     end
 
+    def apply_level3_data!(options, merch_order)
+      options[:merchant_reference] = merch_order.id
+      options[:customer_reference] = merch_order.guest_token
+
+      if merch_order.for_delivery?
+        options[:shipping_address_zip] = merch_order.shipping_address.zipcode
+        options[:shipping_from_zip] = merch_order.shipments.first.stock_location.zipcode
+        options[:shipping_amount] = (merch_order.shipments.first.cost * 100.0).to_i
+      end
+
+      options[:line_items] = []
+
+      merch_order.line_items.each do |li|
+        options[:line_items] << {
+          product_code: li.variant.id,
+          product_description: li.product.description.presence || li.product.name,
+          unit_cost: (li.price * 100.0).to_i,
+          quantity: li.quantity,
+          tax_amount: (li.additional_tax_total * 100.0).to_i,
+          discount_amount: (((merch_order.promo_total.abs.to_f * li.quantity.to_f) / merch_order.quantity.to_f) * 100.0).to_i
+        }
+      end
+
+      merch_order_total = merch_order.total.to_f
```

Comment:
I noticed that here and in the surrounding code you're using `to_f` to convert to Float.

Since floats can suffer from stuff like (**waves hands**) `1.to_f + 1.to_f` equaling `2.0000000001`,
BigDecimal is a safer bet.

If you've already considered that, and have deliberately chosen Floats instead...I'm curious to hear the reasoning.